File: src/vector_search.py
"""
Vector Search - векторный поиск для улучшения системы памяти.
Использует embeddings и similarity search для семантического поиска.
"""
from __future__ import annotations

import json
import numpy as np
from dataclasses import dataclass, asdict
from datetime import datetime
from pathlib import Path
from typing import Any

from .real_tools import ToolResult

try:
    from sentence_transformers import SentenceTransformer
    HAS_SENTENCE_TRANSFORMERS = True
except ImportError:
    HAS_SENTENCE_TRANSFORMERS = False


@dataclass
class VectorDocument:
    """Документ с векторным представлением."""
    doc_id: str
    content: str
    embedding: list[float]
    metadata: dict[str, Any]
    created_at: str


class VectorStore:
    """Векторное хранилище для семантического поиска."""

    def __init__(self, workspace_root: str = ".", model_name: str = "all-MiniLM-L6-v2"):
        self.workspace_root = Path(workspace_root)
        self.vectors_dir = Path.home() / ".claude_code" / "vectors"
        self.vectors_dir.mkdir(parents=True, exist_ok=True)
        self.vectors_file = self.vectors_dir / "vectors.json"

        self.model_name = model_name
        self.model = None
        self.documents: dict[str, VectorDocument] = {}

        if HAS_SENTENCE_TRANSFORMERS:
            try:
                self.model = SentenceTransformer(model_name)
            except Exception as e:
                print(f"⚠️  Warning: Failed to load model {model_name}: {e}")
        else:
            print("⚠️  Warning: sentence-transformers not installed. Vector search disabled!")
            print("   Install with: pip install sentence-transformers")

        self._load_documents()

    def _load_documents(self) -> None:
        """Загрузить документы из файла."""
        if not self.vectors_file.exists():
            return

        try:
            with open(self.vectors_file, 'r', encoding='utf-8') as f:
                data = json.load(f)

            for doc_id, doc_data in data.items():
                self.documents[doc_id] = VectorDocument(**doc_data)

        except Exception as e:
            print(f"Warning: Failed to load vectors: {e}")

    def _save_documents(self) -> None:
        """Сохранить документы в файл."""
        try:
            data = {doc_id: asdict(doc) for doc_id, doc in self.documents.items()}

            with open(self.vectors_file, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)

        except Exception as e:
            print(f"Error saving vectors: {e}")

    def _get_embedding(self, text: str) -> list[float] | None:
        """Получить embedding для текста."""
        if not self.model:
            return None

        try:
            embedding = self.model.encode(text)
            return embedding.tolist()
        except Exception as e:
            print(f"Error getting embedding: {e}")
            return None

    def _cosine_similarity(self, vec1: list[float], vec2: list[float]) -> float:
        """Вычислить косинусное сходство между векторами."""
        v1 = np.array(vec1)
        v2 = np.array(vec2)

        dot_product = np.dot(v1, v2)
        norm1 = np.linalg.norm(v1)
        norm2 = np.linalg.norm(v2)

        if norm1 == 0 or norm2 == 0:
            return 0.0

        return float(dot_product / (norm1 * norm2))

    def add_document(
        self,
        doc_id: str,
        content: str,
        metadata: dict[str, Any] | None = None
    ) -> ToolResult:
        """
        Добавить документ в векторное хранилище.

        Args:
            doc_id: ID документа
            content: Содержимое документа
            metadata: Метаданные

        Returns:
            ToolResult
        """
        if not self.model:
            return ToolResult(
                success=False,
                output="",
                error="Vector search not available (model not loaded)"
            )

        # Получаем embedding
        embedding = self._get_embedding(content)
        if not embedding:
            return ToolResult(
                success=False,
                output="",
                error="Failed to generate embedding"
            )

        # Создаем документ
        document = VectorDocument(
            doc_id=doc_id,
            content=content,
            embedding=embedding,
            metadata=metadata or {},
            created_at=datetime.now().isoformat()
        )

        self.documents[doc_id] = document
        self._save_documents()

        return ToolResult(
            success=True,
            output=f"Document added: {doc_id}",
            error=None
        )

    def search(
        self,
        query: str,
        top_k: int = 5,
        threshold: float = 0.0
    ) -> ToolResult:
        """
        Поиск похожих документов.

        Args:
            query: Поисковый запрос
            top_k: Количество результатов
            threshold: Минимальный порог сходства (0-1)

        Returns:
            ToolResult с результатами поиска
        """
        if not self.model:
            return ToolResult(
                success=False,
                output="",
                error="Vector search not available (model not loaded)"
            )

        if not self.documents:
            return ToolResult(
                success=True,
                output="No documents in vector store",
                error=None
            )

        # Получаем embedding запроса
        query_embedding = self._get_embedding(query)
        if not query_embedding:
            return ToolResult(
                success=False,
                output="",
                error="Failed to generate query embedding"
            )

        # Вычисляем сходство со всеми документами
        results = []
        for doc in self.documents.values():
            similarity = self._cosine_similarity(query_embedding, doc.embedding)

            if similarity >= threshold:
                results.append({
                    "doc_id": doc.doc_id,
                    "content": doc.content,
                    "similarity": similarity,
                    "metadata": doc.metadata
                })

        # Сортируем по убыванию сходства
        results.sort(key=lambda x: x["similarity"], reverse=True)
        results = results[:top_k]

        if not results:
            return ToolResult(
                success=True,
                output="No similar documents found",
                error=None
            )

        # Форматируем результаты
        output = f"Found {len(results)} similar documents:\n\n"
        for i, result in enumerate(results, 1):
            output += f"{i}. [{result['doc_id']}] (similarity: {result['similarity']:.3f})\n"
            output += f"   {result['content'][:200]}\n"
            if result['metadata']:
                output += f"   Metadata: {result['metadata']}\n"
            output += "\n"

        return ToolResult(
            success=True,
            output=output,
            error=None
        )

    def delete_document(self, doc_id: str) -> ToolResult:
        """
        Удалить документ из хранилища.

        Args:
            doc_id: ID документа

        Returns:
            ToolResult
        """
        if doc_id not in self.documents:
            return ToolResult(
                success=False,
                output="",
                error=f"Document not found: {doc_id}"
            )

        del self.documents[doc_id]
        self._save_documents()

        return ToolResult(
            success=True,
            output=f"Document deleted: {doc_id}",
            error=None
        )

    def list_documents(self) -> ToolResult:
        """
        Список всех документов.

        Returns:
            ToolResult со списком документов
        """
        if not self.documents:
            return ToolResult(
                success=True,
                output="No documents in vector store",
                error=None
            )

        output = f"Vector Store Documents ({len(self.documents)}):\n\n"
        for doc in self.documents.values():
            output += f"[{doc.doc_id}]\n"
            output += f"  Content: {doc.content[:100]}...\n"
            output += f"  Created: {doc.created_at}\n"
            if doc.metadata:
                output += f"  Metadata: {doc.metadata}\n"
            output += "\n"

        return ToolResult(
            success=True,
            output=output,
            error=None
        )

    def get_document(self, doc_id: str) -> ToolResult:
        """
        Получить документ по ID.

        Args:
            doc_id: ID документа

        Returns:
            ToolResult с документом
        """
        doc = self.documents.get(doc_id)
        if not doc:
            return ToolResult(
                success=False,
                output="",
                error=f"Document not found: {doc_id}"
            )

        output = f"Document: {doc.doc_id}\n"
        output += f"Content: {doc.content}\n"
        output += f"Created: {doc.created_at}\n"
        output += f"Metadata: {doc.metadata}\n"

        return ToolResult(
            success=True,
            output=output,
            error=None
        )

    def update_document(
        self,
        doc_id: str,
        content: str | None = None,
        metadata: dict[str, Any] | None = None
    ) -> ToolResult:
        """
        Обновить документ.

        Args:
            doc_id: ID документа
            content: Новое содержимое (опционально)
            metadata: Новые метаданные (опционально)

        Returns:
            ToolResult
        """
        doc = self.documents.get(doc_id)
        if not doc:
            return ToolResult(
                success=False,
                output="",
                error=f"Document not found: {doc_id}"
            )

        if content:
            # Пересчитываем embedding
            embedding = self._get_embedding(content)
            if not embedding:
                return ToolResult(
                    success=False,
                    output="",
                    error="Failed to generate embedding"
                )

            doc.content = content
            doc.embedding = embedding

        if metadata:
            doc.metadata.update(metadata)

        self._save_documents()

        return ToolResult(
            success=True,
            output=f"Document updated: {doc_id}",
            error=None
        )

    def clear(self) -> ToolResult:
        """
        Очистить все документы.

        Returns:
            ToolResult
        """
        count = len(self.documents)
        self.documents.clear()
        self._save_documents()

        return ToolResult(
            success=True,
            output=f"Cleared {count} documents from vector store",
            error=None
        )


# Пример использования
if __name__ == "__main__":
    store = VectorStore()

    # Добавляем документы
    store.add_document(
        "doc1",
        "Python is a high-level programming language",
        metadata={"category": "programming"}
    )

    store.add_document(
        "doc2",
        "JavaScript is used for web development",
        metadata={"category": "programming"}
    )

    store.add_document(
        "doc3",
        "Machine learning is a subset of artificial intelligence",
        metadata={"category": "ai"}
    )

    # Поиск
    result = store.search("programming languages", top_k=2)
    print(result.output)

File: src/vector_search.rs
//! Vector Search — семантический поиск по сохранённым документам.
//! Порт `vector_search.py` (итерация 6).
//!
//! ВАЖНОЕ ОТЛИЧИЕ ОТ ОРИГИНАЛА: `vector_search.py` использует
//! `sentence-transformers` (нейросетевые embeddings, модель all-MiniLM-L6-v2,
//! ~90 МБ весов, PyTorch). Тащить PyTorch/ONNX-инференс в чистый Rust —
//! отдельная большая задача (candle + скачивание весов модели), а в этой
//! песочнице нет сети, чтобы даже проверить сборку такой зависимости.
//!
//! Вместо этого здесь честная, отдельно работающая реализация: TF-IDF
//! векторы (частота термина × обратная частота документа) + косинусное
//! сходство. Это классический baseline семантического поиска — хуже
//! настоящих embeddings на перефразированных запросах ("собака" не
//! найдёт "пёс"), но реально работает без единой внешней зависимости
//! и без скачивания модели. Публичный API (`add_document`/`search`/
//! `delete_document`/`list_documents`/`get_document`/`update_document`/
//! `clear`) идентичен оригиналу, так что при желании TF-IDF можно
//! позже заменить на настоящие embeddings (напр. через `candle` или
//! вызов внешнего эмбеддинг-сервера по HTTP через `http_tools.rs`) не
//! трогая вызывающий код.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;

use crate::tools::ToolResult;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VectorDocument {
    pub doc_id: String,
    pub content: String,
    /// TF-IDF вектор в разреженном виде: термин -> вес.
    pub embedding: HashMap<String, f64>,
    pub metadata: HashMap<String, serde_json::Value>,
    pub created_at: String,
}

fn tokenize(text: &str) -> Vec<String> {
    text.to_lowercase()
        .split(|c: char| !c.is_alphanumeric())
        .filter(|w| w.len() > 1)
        .map(|w| w.to_string())
        .collect()
}

pub struct VectorStore {
    vectors_file: PathBuf,
    /// Обратная частота документа (idf) по всем документам в хранилище,
    /// пересчитывается при каждом изменении набора документов — набор
    /// в этом сценарии (память агента) обычно небольшой (десятки-сотни
    /// документов), пересчёт целиком не проблема.
    documents: Mutex<HashMap<String, VectorDocument>>,
}

impl VectorStore {
    pub fn new() -> Self {
        let mut vectors_dir = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
        vectors_dir.push(".hephaestus");
        vectors_dir.push("vectors");
        let _ = fs::create_dir_all(&vectors_dir);
        let vectors_file = vectors_dir.join("vectors.json");
        let documents = Self::load(&vectors_file);
        Self {
            vectors_file,
            documents: Mutex::new(documents),
        }
    }

    fn load(path: &PathBuf) -> HashMap<String, VectorDocument> {
        match fs::read_to_string(path) {
            Ok(s) => serde_json::from_str(&s).unwrap_or_default(),
            Err(_) => HashMap::new(),
        }
    }

    fn save(&self, docs: &HashMap<String, VectorDocument>) {
        if let Ok(json) = serde_json::to_string_pretty(docs) {
            let _ = fs::write(&self.vectors_file, json);
        }
    }

    /// Пересчитать TF-IDF embedding для одного текста относительно
    /// текущего корпуса (используется и при индексации, и при запросе —
    /// запрос не добавляется в корпус, только временно смешивается в idf).
    fn embed(&self, text: &str, corpus: &HashMap<String, VectorDocument>) -> HashMap<String, f64> {
        let tokens = tokenize(text);
        if tokens.is_empty() {
            return HashMap::new();
        }

        let mut tf: HashMap<String, f64> = HashMap::new();
        for t in &tokens {
            *tf.entry(t.clone()).or_insert(0.0) += 1.0;
        }
        let total = tokens.len() as f64;
        for v in tf.values_mut() {
            *v /= total;
        }

        let n_docs = (corpus.len() + 1) as f64; // +1 за сам текст
        let mut result = HashMap::new();
        for (term, tf_val) in tf {
            let df = 1.0 + corpus
                .values()
                .filter(|d| d.content.to_lowercase().contains(&term))
                .count() as f64;
            let idf = (n_docs / df).ln() + 1.0;
            result.insert(term, tf_val * idf);
        }
        result
    }

    fn cosine_similarity(a: &HashMap<String, f64>, b: &HashMap<String, f64>) -> f64 {
        let mut dot = 0.0;
        for (k, va) in a {
            if let Some(vb) = b.get(k) {
                dot += va * vb;
            }
        }
        let norm_a: f64 = a.values().map(|v| v * v).sum::<f64>().sqrt();
        let norm_b: f64 = b.values().map(|v| v * v).sum::<f64>().sqrt();
        if norm_a == 0.0 || norm_b == 0.0 {
            0.0
        } else {
            dot / (norm_a * norm_b)
        }
    }

    pub fn add_document(
        &self,
        doc_id: String,
        content: String,
        metadata: Option<HashMap<String, serde_json::Value>>,
    ) -> ToolResult {
        let mut docs = self.documents.lock().unwrap();
        let embedding = self.embed(&content, &docs);
        docs.insert(
            doc_id.clone(),
            VectorDocument {
                doc_id: doc_id.clone(),
                content,
                embedding,
                metadata: metadata.unwrap_or_default(),
                created_at: chrono::Local::now().to_rfc3339(),
            },
        );
        self.save(&docs);
        ToolResult::success(format!("Document added: {}", doc_id))
    }

    pub fn search(&self, query: &str, top_k: usize, threshold: f64) -> ToolResult {
        let docs = self.documents.lock().unwrap();
        if docs.is_empty() {
            return ToolResult::success("No documents in vector store".to_string());
        }

        let query_embedding = self.embed(query, &docs);
        if query_embedding.is_empty() {
            return ToolResult::success("Query has no indexable terms".to_string());
        }

        let mut results: Vec<(f64, &VectorDocument)> = docs
            .values()
            .map(|d| (Self::cosine_similarity(&query_embedding, &d.embedding), d))
            .filter(|(sim, _)| *sim >= threshold)
            .collect();
        results.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(std::cmp::Ordering::Equal));
        results.truncate(top_k);

        if results.is_empty() {
            return ToolResult::success("No similar documents found".to_string());
        }

        let mut output = format!("Found {} similar documents:\n\n", results.len());
        for (i, (sim, doc)) in results.iter().enumerate() {
            let preview: String = doc.content.chars().take(200).collect();
            output.push_str(&format!("{}. [{}] (similarity: {:.3})\n   {}\n", i + 1, doc.doc_id, sim, preview));
            if !doc.metadata.is_empty() {
                output.push_str(&format!("   Metadata: {}\n", serde_json::Value::Object(doc.metadata.clone().into_iter().collect())));
            }
            output.push('\n');
        }
        ToolResult::success(output)
    }

    pub fn delete_document(&self, doc_id: &str) -> ToolResult {
        let mut docs = self.documents.lock().unwrap();
        if docs.remove(doc_id).is_some() {
            self.save(&docs);
            ToolResult::success(format!("Document deleted: {}", doc_id))
        } else {
            ToolResult::error(format!("Document not found: {}", doc_id))
        }
    }

    pub fn list_documents(&self) -> ToolResult {
        let docs = self.documents.lock().unwrap();
        if docs.is_empty() {
            return ToolResult::success("No documents in vector store".to_string());
        }
        let mut output = format!("Vector Store Documents ({}):\n\n", docs.len());
        for doc in docs.values() {
            let preview: String = doc.content.chars().take(100).collect();
            output.push_str(&format!("[{}]\n  Content: {}...\n  Created: {}\n\n", doc.doc_id, preview, doc.created_at));
        }
        ToolResult::success(output)
    }

    pub fn get_document(&self, doc_id: &str) -> ToolResult {
        let docs = self.documents.lock().unwrap();
        match docs.get(doc_id) {
            Some(doc) => ToolResult::success(format!(
                "Document: {}\nContent: {}\nCreated: {}\n",
                doc.doc_id, doc.content, doc.created_at
            )),
            None => ToolResult::error(format!("Document not found: {}", doc_id)),
        }
    }

    pub fn update_document(
        &self,
        doc_id: &str,
        content: Option<String>,
        metadata: Option<HashMap<String, serde_json::Value>>,
    ) -> ToolResult {
        let mut docs = self.documents.lock().unwrap();
        if !docs.contains_key(doc_id) {
            return ToolResult::error(format!("Document not found: {}", doc_id));
        }
        if let Some(c) = content {
            let embedding = self.embed(&c, &docs);
            if let Some(d) = docs.get_mut(doc_id) {
                d.content = c;
                d.embedding = embedding;
            }
        }
        if let Some(m) = metadata {
            if let Some(d) = docs.get_mut(doc_id) {
                d.metadata.extend(m);
            }
        }
        self.save(&docs);
        ToolResult::success(format!("Document updated: {}", doc_id))
    }

    pub fn clear(&self) -> ToolResult {
        let mut docs = self.documents.lock().unwrap();
        let count = docs.len();
        docs.clear();
        self.save(&docs);
        ToolResult::success(format!("Cleared {} documents from vector store", count))
    }
}

impl Default for VectorStore {
    fn default() -> Self {
        Self::new()
    }
}
